src/runtime/slice.go

```diff
@@ -54,29 +36,21 @@ func sliceCopy(dst, src unsafe.Pointer, dstLen, srcLen uintptr, elemSize uintptr
 
 // sliceGrow returns a new slice with space for at least newCap elements
 func sliceGrow(oldBuf unsafe.Pointer, oldLen, oldCap, newCap, elemSize uintptr) (unsafe.Pointer, uintptr, uintptr) {
-
-	// TODO(dgryski): sliceGrow() and sliceAppend() should be refactored to share the base growth code.
-
 	if oldCap >= newCap {
 		// No need to grow, return the input slice.
 		return oldBuf, oldLen, oldCap
 	}
 
-	// allow nil slice
-	if oldCap == 0 {
-		oldCap++
-	}
-
-	// grow capacity
-	for oldCap < newCap {
-		oldCap *= 2
-	}
+	// This can be made more memory-efficient by multiplying by some other constant, such as 1.5,
+	// which seems to be allowed by the Go language specification (but this can be observed by
+	// programs).
+	newCap = 1 << bits.Len32(uint32(newCap))
```

Comment:
so tinygo is 32bits only? 

also tinygo doesn't need to account for malloc headers like big go?

Reply:
Changed this to use `bits.Len()`, passing an `uint` instead of `uint32`.  Hopefully this should work on 64-bit too.

Reply:
Thanks,

and for the other part (which may not be applicable to tiny go, depends on what kind of heap allocator there is but I think in big go they subtract some meta info size so for instance (making actual numbers up) when requesting 3500 bytes you get a full 4096-headers page of memory instead of 4096+fragment)

ps: I have no idea of tinygo's internal

Reply:
Whether there is metadata or not depends on `-gc=precise` vs `-gc=conservative`. But that should be an implementation detail of the GC/heap: all the slice code needs to know is that there's the `alloc` function and it's not allowed to write outside the returned memory area.